gridt/src/app/core/settings.service.ts

```diff
@@ -22,27 +24,98 @@ export class SettingsService {
+  * identity of the user.
+  */
+  public userIdentity$: Observable<Identity> = this._user_identity$.asObservable();
+
+  /**
+  * This function updates the user identity.
+  * - If both server and local identity are unavailable, emit empty ID.
+  * - When server ID is retreived, but local ID has not been set yet, store
+  *   server ID in local and emit local ID.
+  * - When the server ID is different from local ID, overwrite local ID with
+  *   the server ID. Then emit local ID.
+  * - If both the server and local identity agree, emit local ID.
+  * - If the server is unavailable but local ID is available, emit local ID.
+  * - Only when updateIdentity() function is called should the userIdentity$
+  *   change.
+  *
+  * - Makes sure all errors are caught such that this function always gives an
+  *   output.
+  */
+  private mapErrorToEmptyID = pipe(
+    catchError( err => {console.warn(err); return of<Identity>(this.empty_identity);}),
+    flatMap( (id) => of(<Identity>(id)))
+  );
+
+  private mapErrorToFalse = pipe(
+    catchError( err => {console.warn(err); return of<boolean>(false);})
+  );
+
+  public updateIdentity(): void {
+    // Combine server and local id. Catch errors to make sure pipe runs.
+    forkJoin({
+      server: this.api.userIdentity$().pipe( this.mapErrorToEmptyID ),
+      local: this.localIdentity$.pipe( this.mapErrorToEmptyID )
+    }).pipe(
+      // Update local id when server id is different.
+      tap( (ids) => console.debug(`recieved server id = ${JSON.stringify(ids.server)} \n received local id = ${JSON.stringify(ids.local)}`)),
+      flatMap( (ids) => iif(
+        () => ((JSON.stringify(ids.server)==JSON.stringify(ids.local)) || (JSON.stringify(ids.server) == JSON.stringify(this.empty_identity))), // && ids[0].id == ids[1].id
+        of(true),
+        this.setLocalIdentity$(ids.server)
+      )),
+      // Make sure any errors arising from setting local identity are
+      // transformed into a warning and continue.
+      this.mapErrorToFalse,
+      // Emit local identity
+      flatMap( () => this.localIdentity$ ),
+      // If local identity unvailable emit empty ID.
+      this.mapErrorToEmptyID,
+      // make sure subscription completes and gives only one update.
+      take(1)
+    ).subscribe( (id) => this._user_identity$.next(id));
+  };
+
   /**
-   * Observable to obtain settings from secure storage
+   * Observable to obtain identity from secure storage
    * If secure storage is empty, this returns an unavailable error
    */
   public localIdentity$: Observable<Identity> = this.auth.readyAuthentication$.pipe(
     catchError( () => throwError(this.error_codes.GETIDFAIL + ": " + this.error_codes.NOTLOGGEDIN)),
     flatMap(() => this.secStore.get$("identity").pipe(
       catchError( () => throwError(this.error_codes.GETIDFAIL + ": " + this.error_codes.SECSTOREUNAVAILABLE))
-    ))
+    )),
+    flatMap( (id) => of<Identity>(JSON.parse(id)))
```

Comment:
As far as I understand, the output of secstore is already parsed. Why are you parsing it again?

Reply:
No?? I don't think so, here is a copy of the secStore.get$ function
```
get$ (key: string): Observable<any> {
    return new Observable ( (observer) => {
      SecureStoragePlugin.get({key}).then(
        (valueObj: any) => {
          const value = valueObj.value;
          if (value === atob(null) || !value){
            observer.error(`Key "${key}" does not exist in the secure storage.`);
            return;
          } else {
            observer.next(value);
            observer.complete();
          }
        }
      ).catch( () => observer.error(`Key "${key}" does not exist in the secure storage.`) );
    });
```
This simply outputs the value as was stored in the secStore.

Reply:
Oh dear, this is a problem, I have added that functionality some where, maybe it is still in an active pull request?

Reply:
Remember that time my draft pulled into development? I thought I fixed it by just reverting the changes. Turns out there where some changes made after that mistake. Guess what... That was your secStore change. So unfortunately those changes got overwritten... Not sure how to continue, should we accept this PR, open a new issue and include the changes you made earlier. Or shall I include your changes in this PR?